src/sgame/lua/Interpreter.cpp

```diff
@@ -0,0 +1,321 @@
+
+using Shared::Lua::LuaLib;
+
+namespace Lua {
+
+static lua_State* L = nullptr;
+
+bool LoadCode( Str::StringRef code, Str::StringRef location )
+{
+	if ( luaL_loadbuffer( L, code.c_str(), code.size(), location.c_str() ) != 0 )
+	{
+		Shared::Lua::Report( L, "Loading buffer" );
+		return false;
+	}
+	return true;
+}
+
+bool RunCode()
+{
+	if ( lua_pcall( L, 0, 0, 0 ) != 0 )
+	{
+		Shared::Lua::Report( L, "Executing code" );
+		return false;
+	}
+	return true;
+}
+
+bool LoadScript( Str::StringRef scriptPath )
+{
+	fileHandle_t f;
+	int len = trap_FS_FOpenFile( scriptPath.c_str(), &f, fsMode_t::FS_READ );
+	if ( len <= 0 )
+	{
+		Log::Warn( "error opening %s: %d", scriptPath, len );
+		return false;
+	}
+	std::vector<char> code( len + 1 );
+	code[ len ] = '\0';
+	int ret = trap_FS_Read( code.data(), len, f );
+	if ( ret != len )
+	{
+		Log::Warn( "error reading %s: %d", scriptPath, ret );
+		return false;
+	}
+	return LoadCode( code.data(), scriptPath );
+}
+
+/*
+ * Below here are global functions and their helper functions that help overwrite the Lua global functions. The majority of this file was copied from LibRocket.
+ */
+
+// __pairs should return two values.
+// upvalue 1 is the __pairs function, upvalue 2 is the userdata created in rocket_pairs
+// [1] is the object implementing __pairs
+// [2] is the key that was just read
+int rocket_pairsaux( lua_State* L )
+{
+	lua_pushvalue( L, lua_upvalueindex( 1 ) );  // push __pairs to top
+	lua_insert( L, 1 );                         // move __pairs to the bottom
+	lua_pushvalue( L, lua_upvalueindex( 2 ) );  // push userdata
+	// stack looks like [1] = __pairs, [2] = object, [3] = latest key, [4] = userdata
+	if ( lua_pcall( L, lua_gettop( L ) - 1, LUA_MULTRET, 0 ) != 0 )
+		Shared::Lua::Report( L, "__pairs" );
+	return lua_gettop( L );
+}
+
+// A version of pairs that respects a __pairs metamethod.
+// "next" function is upvalue 1
+int rocket_pairs( lua_State* L )
+{
+	luaL_checkany( L, 1 );  // [1] is the object given to us by pairs(object)
+	if ( luaL_getmetafield( L, 1, "__pairs" ) )
+	{
+		void* ud = lua_newuserdata(
+			L, sizeof( void* ) );  // create a new block of memory to be used as upvalue 1
+		( *(int*)( ud ) ) = -1;
+		lua_pushcclosure( L, rocket_pairsaux, 2 );  // uv 1 is __pairs, uv 2 is ud
+	}
+	else
+	{
+		lua_pushvalue( L, lua_upvalueindex( 1 ) );  // generator
+	}
+	lua_pushvalue( L, 1 );  // state
+	lua_pushnil( L );       // initial value
+	return 3;
+}
+
+// copy + pasted from Lua's lbaselib.c
+int ipairsaux( lua_State* L )
+{
+	int i = luaL_checkinteger( L, 2 );
+	luaL_checktype( L, 1, LUA_TTABLE );
+	i++; /* next value */
+	lua_pushinteger( L, i );
+	lua_rawgeti( L, 1, i );
+	return ( lua_isnil( L, -1 ) ) ? 0 : 2;
+}
+
+// __ipairs should return two values
+// upvalue 1 is the __ipairs function, upvalue 2 is the userdata created in rocket_ipairs
+// [1] is the object implementing __ipairs, [2] is the key last used
+int rocket_ipairsaux( lua_State* L )
```

Comment:
As the code is written, removing this code makes the existing code not work.

It was removed, but only because it was changed. https://github.com/mikke89/RmlUi/pull/137

I'd like to clean this up as a follow up PR to this one since I need to research their changes. It's clear that this is bad when integrating with third party libraries (and with the cgame :cry: , so we should probably clean this up.

Reply:
> As the code is written, removing this code makes the existing code not work.

Which existing code?

Reply:
The iterators when you do:
```
for k, v in pairs(Unv.weapons) do
  print(k, v)
end
```

Same happens for buildables, classes, upgrades, etc

Reply:
Maybe `weapons` should just be a table then? Then everything will just work as expected with the same interface without a ton of custom code. In the merged code so far `Player` was the only class that couldn't have been replaced with a plain table. The only caveat I know of is that users could modify the table

Reply:
There's a [solution to make a table immutable](https://www.lua-users.org/wiki/ReadOnlyTables)

Reply:
It basically is a table. I'll just fix the iterator stuff in another PR.

Reply:
https://github.com/Unvanquished/Unvanquished/pull/3033